### DISCUSSION TITLE
Multiple input files, same operation

### DISCUSSION THREAD

**Discussion by Mastomaki** [Spine-project/Spine-Toolbox#2924] [Support]
If I want to perform the same operation on multiple input files, do I create multiple tools for that with the same code? Or is the possible to use the same tool for multiple input files, one at the time?

**Comment by soininen** [comment]
One solution would be to use a single Data Connection that provides the files to the tool:
![image](https://github.com/user-attachments/assets/4352f553-022b-46dc-ba93-41a08f121f91)

In the tool specification, you can request the files as Optional input files using wildcards:
![image](https://github.com/user-attachments/assets/c0cab4b3-d6b7-4b9c-87d7-4a15733e140c)

Toolbox will copy the file references from the Data connection that match the Optional input files into the tool's work directory before the tool is executed. You can then access the files from the tool script as if they were in the current directory.

**Comment by Mastomaki** [comment]
In this case is the tool executed as many times are there are input files?

**Reply by soininen** [reply]
Is your tool somehow restricted to processing a single file at a time? In the above example, the script is executed only a single time but it processes all files in the `for` loop.

**Reply by Mastomaki** [reply]
Yes it processes a single file. If it should process many files, then each of them becomes a command line argument?

**Reply by soininen** [reply]
If your tool processes just a single file, then the basic approach would indeed be to have a Data connection and a Tool for each file:
![image](https://github.com/user-attachments/assets/721c9383-1497-4e60-9600-6a77c72f946a)

Luckily, the Tools can share the same specification, so modifying one will modify all of them. We do not have a mechanism to "distribute" input files to different instances of the same tool, except perhaps by using Loops but that is rather difficult to set up.

Would it be possible to wrap your tool in a Python script that loops over the input files and launches the tool for each of them? Something like the script in the example above?

> If it should process many files, then each of them becomes a command line argument?

It is one possibility but then you need to manage the command line arguments manually which may be an issue if you have to do it often. In the above Python script example, I have used Optional input files instead of command line arguments so I do not need to modify them if e.g. the number of input files change.
